libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -451,6 +478,144 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
   return 0;
 }
 
+bool VoteManager::addRewardVote(const std::shared_ptr<Vote>& vote) {
+  const auto vote_hash = vote->getHash();
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_dg_) << "Drop vote " << vote_hash << ". PBFT chain last block is " << pbft_chain_last_block_hash
+                 << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  std::unique_lock lock(reward_votes_mutex_);
+  if (reward_votes_.contains(vote_hash)) {
+    LOG(log_dg_) << "Reward vote " << vote_hash << " saved already";
+    return false;
+  }
+
+  if (reward_votes_pbft_block_hash_ != voted_block_hash) {
+    LOG(log_er_) << "The reward vote voted PBFT block hash " << voted_block_hash << " is different with "
+                 << reward_votes_pbft_block_hash_;
+    assert(false);
+  }
+
+  if (!verifyRewardVote(vote)) return false;
+
+  reward_votes_.insert({vote_hash, vote});
+  db_->saveLastBlockCertVote(vote);
+  LOG(log_nf_) << "add reward vote " << vote_hash;
+
+  return true;
+}
+
+bool VoteManager::verifyRewardVote(const std::shared_ptr<Vote>& vote) {
```

Comment:
do we need to run for already verified votes?

Reply:
In VotePacketHandler::process, we should never verify any already verified votes, since we check seen votes before addRewardVote.
For move votes in verified_votes_ to reward_votes_, since PBFT has moved to a new round, PBFT may finalize new PBFT blocks and that may change DPOS states. For those votes, need verify again.